cmd/opampsupervisor/supervisor/packages.go

```diff
@@ -0,0 +1,412 @@
+		}, nil
+	}
+
+	return types.PackageState{
+		Exists: false,
+	}, nil
+}
+
+func (p *packageManager) SetPackageState(packageName string, state types.PackageState) error {
+	if packageName != agentPackageKey {
+		return fmt.Errorf("package %q does not exist", packageName)
+	}
+
+	if !state.Exists {
+		return fmt.Errorf("agent package must be marked as existing")
+	}
+
+	if state.Type != protobufs.PackageType_PackageType_TopLevel {
+		return fmt.Errorf("agent package must be marked as top level")
+	}
+
+	p.topLevelHash = state.Hash
+	p.topLevelVersion = state.Version
+
+	return nil
+}
+
+func (packageManager) CreatePackage(packageName string, _ protobufs.PackageType) error {
+	if packageName != agentPackageKey {
+		return fmt.Errorf("only agent package is supported")
+	}
+
+	return fmt.Errorf("agent package already exists")
+}
+
+func (p *packageManager) FileContentHash(packageName string) ([]byte, error) {
+	if packageName != agentPackageKey {
+		return nil, nil
+	}
+
+	return p.topLevelHash, nil
+}
+
+func (p *packageManager) UpdateContent(ctx context.Context, packageName string, data io.Reader, contentHash, signature []byte) error {
+	if packageName != agentPackageKey {
+		return fmt.Errorf("package does not exist")
+	}
+
+	by, err := io.ReadAll(data)
+	if err != nil {
+		return fmt.Errorf("read package bytes: %w", err)
+	}
+
+	err = verifyPackageHash(by, contentHash)
+	if err != nil {
+		return fmt.Errorf("could not verify package integrity: %w", err)
+	}
+
+	b64Cert, b64Signature, err := parsePackageSignature(signature)
+	if err != nil {
+		return fmt.Errorf("could not parse package signature: %w", err)
+	}
+
+	err = verifyPackageSignature(ctx, p.checkOpts, by, b64Cert, b64Signature)
+	if err != nil {
+		return fmt.Errorf("could not verify package signature: %w", err)
+	}
+
+	// overwrite agent process
+	startAgent, err := p.am.stopAgentProcess(ctx)
+	if err != nil {
+		return fmt.Errorf("stop agent process: %w", err)
+	}
+
+	// We always want to start the agent process again, even if we fail to write the agent file
+	defer close(startAgent)
+
+	// Create a backup in case we fail to write the agent
+	// verify collector backup path is clear
+	agentBackupPath := filepath.Join(p.storageDir, "collector.bak")
+	if err = renameFile(p.agentExePath, agentBackupPath); err != nil {
+		return fmt.Errorf("rename collector exe path to backup path: %w", err)
+	}
+
+	// Create reader for new agent
+	gzipReader, err := gzip.NewReader(bytes.NewBuffer(by))
+	if err != nil {
+		return fmt.Errorf("create gzip reader: %w", err)
+	}
+	defer gzipReader.Close()
+
+	tar := tar.NewReader(gzipReader)
+	h, err := tar.Next()
+	if err != nil {
+		return fmt.Errorf("read tarball for collector: %w", err)
+	}
+
+	for h.Name != "otelcol-contrib" {
+		h, err = tar.Next()
+		if err != nil {
+			return fmt.Errorf("read tarball for collector: %w", err)
+		}
+	}
+
+	// open collector destination file
+	agentFile, err := os.OpenFile(p.agentExePath, os.O_RDWR|os.O_CREATE, 0o700)
```

Comment:
What do you think of a slightly different sequence of operations when updating:

- Download the tarbal
- Extract the tarbal to a temp file in the same dir as the agentExePath
- Stop the agent
- Rename agentExePath to agentBackupPath
- Rename temp file to agentExePath
- Start the agent
- Delete backup

This way we don't stop the agent until we are certain that the tarball can be extracted and written to disk, so slightly reduced chance that it will fail (e.g. because of low disk space).